### PR TITLE
[yul-phaser] Miscellaneous Population tweaks and improvements

### PR DESCRIPTION
**Pull request by cameel**
### Description
The fourth pull request implementing #7806. Originally a part of #8256 which is now closed. Depends on #8324.

This is a bunch of loosely related changes that are used by later PRs but are not really a part of the "business logic" of the application. Extracted into a separate PR to make the others smaller and easier to review.

1. `Chromosome` can now be constructed from a string with step abbreviations.
2. Operator `+` for `Population`. This will be a core mechanic in the algorithms as the `Population` becomes immutable in one of the next PRs (#8327) but is also very convenient in general.
3. Ability to specify desired chromosome length in `Population::makeRandom()`.
4. Deterministic order of individuals. Now it takes into account the chromosome so that individuals with the same fitness are always ordered the same way (lexicographically). A side-benefit is that the shorter chromosomes are now preferred.
5. Equality operators for `Population` and `Individual`.
6. More helpers added to `test/yulPhaser/Common.h`
7. Fixture that takes the repetitive program construction out of the test cases.
8. Improved tests for `makeRandom()` based on the improvements from the `SimulationRNG` PR (#8324).

Some of the new features (equality operators, `Chromosome` constructor, etc.) are not yet necessary in non-test code but they simplify tests **a lot**. Especially the constructor removes the need for a long list of extra includes. I considered putting them in `Common.h` instead (and in some cases I did - there are some new helpers there) but in my opinion they're not out of place in the normal code and might come in handy later. Being able to easily construct and compare objects does make for a better class API in my opinion.

### Dependencies
This PR is based on #8324. Unfortunately changes from that base PR will show through in the combined diff and on the commit list until it gets merged.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### CODE REVIEW

File: test/yulPhaser/Chromosome.cpp [hunk 1]
```diff
@@ -39,6 +49,24 @@ namespace solidity::phaser::test
...
+		ForLoopConditionOutOfBody::name,
+		Rematerialiser::name,
+		ForLoopConditionOutOfBody::name,
+		ExpressionSimplifier::name,
+		ForLoopInitRewriter::name,
+		LoopInvariantCodeMotion::name,
+		ExpressionInliner::name
+	};
+
+	BOOST_TEST(Chromosome("ChrOmOsoMe").optimisationSteps() == expectedSteps);
```
Comment (chriseth) [inline test/yulPhaser/Chromosome.cpp:67]: :)

**Comment by chriseth** [comment]
Looks good! Please rebase.

**Comment by chriseth** [comment]
Ah, there is a build failure on windows:
```
C:\projects\solidity\test\yulPhaser\Population.cpp(147): error C3493: 'maxLength' cannot be implicitly captured because no default capture mode has been specified [C:\projects\solidity\build\test\soltest.vcxproj]
C:\projects\solidity\test\yulPhaser\Population.cpp(148): error C2665: 'solidity::phaser::Population::makeRandom': none of the 2 overloads could convert all the argument types [C:\projects\solidity\build\test\soltest.vcxproj]
  C:\projects\solidity\tools/yulPhaser/Population.h(78): note: could be 'solidity::phaser::Population solidity::phaser::Population::makeRandom(solidity::phaser::Program,size_t,std::function<size_t (void)>)'
  C:\projects\solidity\test\yulPhaser\Population.cpp(148): note: while trying to match the argument list '(solidity::phaser::Program, const size_t, int)'
C:\projects\solidity\test\yulPhaser\Population.cpp(159): error C3536: 'population': cannot be used before it is initialized [C:\projects\solidity\build\test\soltest.vcxproj]
C:\projects\solidity\test\yulPhaser\Population.cpp(153): error C2672: 'count_if': no matching overloaded function found [C:\projects\solidity\build\test\soltest.vcxproj]
C:\projects\solidity\test\yulPhaser\Population.cpp(159): error C2780: 'iterator_traits<_FwdIt>::difference_type std::count_if(_ExPo &&,const _FwdIt,const _FwdIt,_Pr) noexcept': expects 4 arguments - 2 provided [C:\projects\solidity\build\test\soltest.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\algorithm(291): note: see declaration of 'std::count_if'
C:\projects\solidity\test\yulPhaser\Population.cpp(159): error C2780: 'iterator_traits<_Iter>::difference_type std::count_if(_InIt,_InIt,_Pr)': expects 3 arguments - 2 provided [C:\projects\solidity\build\test\soltest.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\algorithm(269): note: see declaration of 'std::count_if'
```

**Comment by cameel** [comment]
Rebased.

Yeah, looks like some of these PRs are failing 2-3 checks. When I was pushing the code during the weekend there were some totally unrelated failures (e.g. being unable to download some system packages during one of the setup steps due to HTTP 404 from the repo) so I decided to deal with them later. I'll fix them today.

**Review by chriseth — DISMISSED** [review]

File: test/yulPhaser/Chromosome.cpp (hunk 1 above)
Comment (cameel) [reply test/yulPhaser/Chromosome.cpp:67]: I wanted to do `soLiDiTy` but there's no `y` :(

**Comment by cameel** [comment]
I did not have time to do it yesterday after all but now the fix is ready.

I think that the only real error here is the first one (`'maxLength' cannot be implicitly captured because no default capture mode has been specified`) and the others look like collateral damage caused by the missing declaration. But I can't be sure without compiling it so we'll see what the CI says.

The `maxLength` error itself seems to be a known MSVC bug: [Lambda fails to implicitly capture constexpr value: error C3493](https://developercommunity.visualstudio.com/content/problem/610504/lambda-fails-to-implicitly-capture-constexpr-value.html). The variable is `constexpr` and should not have to be listed as a capture but MSVC does not agree. GCC on the other hand issues a warning to point out it's superfluous if I do capture it so I can't do it that way either. Well, fortunately `maxLength` is just a local integer variable and I can live with it not being `constexpr`. I simply removed the qualifier in this particular function and also in a similar spot in #8328 to make both compilers happy.

**Comment by cameel** [comment]
All the tests passed. `continuous-integration/appveyor/pr` is still marked as pending here for some reason but if you look at the details it says that the build has succeeded.

**Review by chriseth — APPROVED** [review]

**Pull request merged**